generic_tracers/generic_COBALT.F90

```diff
@@ -3934,116 +4004,157 @@ subroutine generic_COBALT_update_from_source(tracer_list,Temp,Salt,rho_dzt,dzt,h
+    ! minerals is available to be remineralized.  The "unprotected" organic fraction is calculated as:
+    !
+    ! max( 0.0,f_ndet - rpcaco3*(cadet_arag+cadet_calc) - rplith*lithdet - rpsio2*sidet)
+    !
+    ! Where rpcaco3, rplith and rpsio2 are protection factors associated with each mineral (Dunne et al., 2005).
+    !
+    ! As was the case for free-living bacteria, the remineralization rate for sinking detritus under anaerobic
+    ! conditions is scaled by o2_min/(k_o2+o2_min).  All anaerobic remineralization is assumed to occur via
+    ! denitrification, so a scaling
+    !
+    ! References:
+    ! Laufkotter et al., 2017: https://agupubs.onlinelibrary.wiley.com/doi/full/10.1002/2017GB005643
+    ! Armstrong, 2002: https://doi.org/10.1016/S0967-0645(01)00101-1
+    ! Klaas and Archer, 2002: https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2001GB001765
+    ! Dunne et al., 2005: https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2004GB002390    
     !
-
-    ! Calculate the depth for scaling of remineralization near the surface
-    allocate(z_remin_ramp(isd:ied,jsd:jed,1:nk)); z_remin_ramp = 0.0
-    z_remin_ramp(:,:,1) = dzt(:,:,1)
-    do k = 2,nk !{
-      z_remin_ramp(:,:,k) = z_remin_ramp(:,:,k-1) + dzt(:,:,k)
-    enddo !}k
-!
-!---------------------------------------------------------------------------------------------------------
-
     do k=1,nk ; do j=jsc,jec ; do i=isc,iec  !{
        cobalt%expkreminT(i,j,k) = exp(cobalt%kappa_remin * Temp(i,j,k))
-       !cobalt%jno3denit_wc(i,j,k) = 0.0
-       !
-       !   Under oxic conditions
-       !
+       ! Calculate remineralization under aerobic remineralization
        if (cobalt%f_o2(i,j,k) .gt. cobalt%o2_min) then  !{
           cobalt%jremin_ndet(i,j,k) = cobalt%gamma_ndet * cobalt%expkreminT(i,j,k) * &
-               z_remin_ramp(i,j,k)/(z_remin_ramp(i,j,k) + cobalt%remin_ramp_scale) * cobalt%f_o2(i,j,k) / &
+               zbot(i,j,k)/(zbot(i,j,k) + cobalt%remin_ramp_scale) * cobalt%f_o2(i,j,k) / &
                ( cobalt%k_o2 + cobalt%f_o2(i,j,k) )*max( 0.0, cobalt%f_ndet(i,j,k) - &
                cobalt%rpcaco3*(cobalt%f_cadet_arag(i,j,k) + cobalt%f_cadet_calc(i,j,k)) - &
                cobalt%rplith*cobalt%f_lithdet(i,j,k) - cobalt%rpsio2*cobalt%f_sidet(i,j,k) )
+          ! Augment total nh4 production and o2 consumption
           cobalt%jprod_nh4(i,j,k) = cobalt%jprod_nh4(i,j,k) + cobalt%jremin_ndet(i,j,k)
           cobalt%jo2resp_wc(i,j,k) = cobalt%jo2resp_wc(i,j,k) + cobalt%jremin_ndet(i,j,k)*cobalt%o2_2_nh4
-       !
-       ! Under sub-oxic conditions
-       !
+
+       ! Calculate remineralization under anaerobic conditions
        else !}{
           cobalt%jremin_ndet(i,j,k) = cobalt%gamma_ndet * cobalt%o2_min / &
                (cobalt%k_o2 + cobalt%o2_min)* &
-               cobalt%f_no3(i,j,k) / (phyto(SMALL)%k_no3 + cobalt%f_no3(i,j,k))* &
+               cobalt%f_no3(i,j,k) / (cobalt%k_no3_denit + cobalt%f_no3(i,j,k))* &
```

Comment:
The new code makes more sense to me, but it also changes our baseline in the NWA12-RT case. This is another example of how a 1D case cannot detect this kind of change.

Reply:
Does it?  I think cobalt(SMALL)%k_no3 and cobalt%k_no3_denit have the same value.  We just introduced k_no3_denit to be explicit that this was a different paramater.

Reply:
I stand corrected.  These parameters are different.  This will change answers, but in a minimal way.  The parameter is only has a significant effect when both oxygen and nitrate have been exhausted which rarely, if ever, happens.  Even in such cases, the impact will be small.

In general, I suggest we remove uses of the phytoplankton parameters beyond their explicit definition.  It looks like we used phyto(SMALL)%k_no3, for example, for the anammox dependence on NO3.  Thus, changing phyto(SMALL)%k_no3 risks unintentionally changing multiple factors.  I think we should adopt k_no3_denit for denitrification and define k_no2_anammox for anammox and set both to 1e-6.  What do you folks think?

@yichengt900 @jessluo

Reply:
Yes, I completely agree. It was just a quick shorthand to use phyto(SMALL)%k_no3 but doing so poses risks where a user may come in and change what they think is just the small phytoplankton nitrate half saturation constant and end up changing values for other processes as well.